### DISCUSSION TITLE
41999 + remote improvements and 42124  (Bybricks beta only)

### DISCUSSION THREAD

**Discussion by Gcopper1984** [pybricks/support#399] [Show and tell]
Hi everyone,
   I just made a new script to use the remote controller and control+ for 42124 (buggy) and 42099 (crawler).

Controller are more or less the same:

Left - and Left + to steer left and right
Left red button to BRAKE

Right + and Right - to drive forward or backward at 100% power. Hold the Right center button with R+ or R- to drive at 50% of power.

Also for 42099 the led color changes from green, orange and red if inclination angle increase.

Hope could be useful for someone.

Here code:

[20210716_42099_Pybricks_v3_DC_Colori.py.txt](https://github.com/pybricks/support/files/6858416/20210716_42099_Pybricks_v3_DC_Colori.py.txt)
[20210716_42124_buggy_RC.py.txt](https://github.com/pybricks/support/files/6858418/20210716_42124_buggy_RC.py.txt)

Hope this help. Any feedback is welcome

GC

**Comment by dlech** [comment]
Here are the same files so people can see them without having download them if they don't want to:

**20210716_42099_Pybricks_v3_DC_Colori.py.txt**

```python
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor, Remote
from pybricks.parameters import Port, Direction, Stop, Button, Color
from pybricks.tools import wait

# Initialize the hub.
hub = TechnicHub()
status = 0 # 0: OK, 1 Warning, 2: Error

# Initialize the motors.
steer = Motor(Port.C)
front = Motor(Port.A, Direction.COUNTERCLOCKWISE)
rear = Motor(Port.B, Direction.COUNTERCLOCKWISE)

# Lower the acceleration so the car starts and stops realistically.
front.control.limits(acceleration=2000) #1000
rear.control.limits(acceleration=2000) #1000

# Connect to the remote.
remote = Remote()
#print(remote.address()) Not supported yet
#remote.light.on(Color.CYAN)

# Find the steering endpoint on the left and right.
# The middle is in between.
steer.reset_angle(0)
left_end = steer.run_until_stalled(-300, then=Stop.HOLD) # 200
right_end = steer.run_until_stalled(300, then=Stop.HOLD)

#print("Left: ", left_end)
#print("Right: ", right_end)

# We are now at the right. Reset this angle to be half the difference.
# That puts zero in the middle.
steer.reset_angle((right_end - left_end) / 2)
steer.run_target(speed=300, target_angle=0, wait=False)

# Now we can start driving!
while True:
    # LED indicating particular situation
    pitch, roll = hub.imu.tilt()
    #print(pitch, roll)
    #wait(1000)

    status = 0

    if pitch < -10 or roll < -10 or pitch > 10 or roll > 10:
        status = 1

    if pitch < -20 or roll < -20 or pitch > 20 or roll > 20:
        status = 2
    
    if status == 0:
        hub.light.on(Color.GREEN)
    elif status == 1:
        hub.light.on(Color.ORANGE)
    elif status == 2:
        hub.light.on(Color.RED)

    # Check which buttons are pressed.
    pressed = remote.buttons.pressed()
    #print("pressed: ", pressed)

    # Choose the steer angle based on the left controls.
    steer_angle = 0
    if Button.LEFT_PLUS in pressed:
        steer_angle -= 90 # 75
    if Button.LEFT_MINUS in pressed:
        steer_angle += 90

    # Steer to the selected angle.
    steer.run_target(500, steer_angle, wait=False)

    # Choose the drive speed based on the right controls.
    drive_speed = 0
    if Button.RIGHT_PLUS in pressed:
        drive_speed += 100 # 1000
    if Button.RIGHT_MINUS in pressed:
        drive_speed -= 100 # 1000 

    # HOLD if left is pressed, else apply the selected speed.
    if Button.LEFT in pressed:
        front.hold()
        rear.hold()
    elif Button.RIGHT in pressed:
        front.dc(drive_speed / 2)
        rear.dc(drive_speed / 2)
    else:
        front.dc(drive_speed)
        rear.dc(drive_speed)

    # Wait.
    wait(10)
```

**20210716_42124_buggy_RC.py.txt**

```python
from pybricks.pupdevices import Motor, Remote
from pybricks.parameters import Port, Direction, Stop, Button
from pybricks.tools import wait

# Initialize the motors.
steer = Motor(Port.B, Direction.COUNTERCLOCKWISE)
drive = Motor(Port.A, Direction.COUNTERCLOCKWISE)

# Lower the acceleration so the car starts and stops realistically.
drive.control.limits(acceleration=1000) #1000

# Connect to the remote.
remote = Remote()
#print(remote.address()) Not supported yet
#remote.light.on(Color.CYAN)

# Find the steering endpoint on the left and right.
# The middle is in between.
steer.reset_angle(0)
left_end = steer.run_until_stalled(-300, then=Stop.HOLD) # 200
right_end = steer.run_until_stalled(300, then=Stop.HOLD)

print("Left: ", left_end)
print("Right: ", right_end)

# We are now at the right. Reset this angle to be half the difference.
# That puts zero in the middle.
steer.reset_angle((right_end - left_end) / 2)
steer.run_target(speed=300, target_angle=0, wait=False)

# Now we can start driving!
while True:
    # Check which buttons are pressed.
    pressed = remote.buttons.pressed()
    #print("pressed: ", pressed)

    # Choose the steer angle based on the left controls.
    steer_angle = 0
    if Button.LEFT_PLUS in pressed:
        steer_angle -= 90 # 75
    if Button.LEFT_MINUS in pressed:
        steer_angle += 90

    # Steer to the selected angle.
    steer.run_target(500, steer_angle, wait=False)

    # Choose the drive speed based on the right controls.
    drive_speed = 0
    if Button.RIGHT_PLUS in pressed:
        drive_speed += 100 # 1000
    if Button.RIGHT_MINUS in pressed:
        drive_speed -= 100 # 1000 

    # HOLD if left is pressed, else apply the selected speed.
    if Button.LEFT in pressed:
        drive.hold()
    elif Button.RIGHT in pressed:
        #drive.run(drive_speed / 2)
        drive.dc(drive_speed / 2)
    else:
        drive.dc(drive_speed)

    # Wait.
    wait(10)
```

**Reply by laurensvalk** [reply]
Thanks for sharing this! Are you `giobaxNXT` who commented about this on our YouTube by any chance?

We'd love to share it on our site with your permission (and credit). It would look similar to [this project](https://pybricks.com/projects/sets/technic/42099-off-roader/powered-up-remote/). What do you think?

Before we do, here are a few tips that might improve the code:
- You used `dc` instead of `run`. This is totally fine, but maybe you can change these things accordingly:
   - You can delete the line with `drive.control.limits`, since `dc` does not use acceleration.
   - Rename `drive_speed` to something like `drive_power` or `duty_cycle`.
- It would be good to delete commands that have been commented out, like `#drive.run(drive_speed / 2)` and `#print("pressed: ", pressed)` or `#200`.

**Reply by laurensvalk** [reply]
Coincidentally, @jorgepe has also just shared his example in a pull request. Perhaps we can merge his pull request, and then @Gcopper1984 can add additional things like the lights if you like?

We can update the template to allow multiple authors, which would be a good thing to do anyway.

**Reply by Gcopper1984** [reply]
Hi!

Yep, it's me!


Ok, thank you for your feedback. Of @JorgePe it's working on that and already proposed a pull request no problem. I Will eventually modifiy his code.

I Hope this project Will grow in future because ot has a lot.of potential.
I Will try to add something.

Thank you!

**Reply by JorgePe** [reply]
Hi.
To be honest, there is no such thing as "my code".  I had the 42124 with me, last day of holidays with the kids and didn't want them to go outside with an iPad just to drive a LEGO car so I simply copied the code from other project (the Technic off-road truck), removed the code for one of the motors and corrected the motor ports. It worked so goal accomplished let's wait for the kids to wake up (later on I found out the calibration of the steering could be better).

Then I realized I could create a new Project just by "copy&paste", replacing links and photos and making a video so why not?

If you @Gcopper1984 have your own code, it would be great if you replace "mine" has we already have a placeholder available.

**Comment by Danikto47** [comment]
Hello.
I want to add light to my buggy 42124. I would be very grateful for your help in finalizing the code so that another port (C or D) can be used to turn on and off the light with a button on the remote control.
Thank you in advance!

**Reply by Gcopper1984** [reply]
Hi,

 just posted this one:

#https://github.com/pybricks/support/discussions/504

Scroll discussion to the end and you will find the code. Led light is attached to port C
